bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/configurable/ConfigTomlParser.java

```diff
@@ -122,105 +127,129 @@ private static Object validateNodeAndExtractValue(VariableKey key, Map<String, T
-                tomlValue = valueMap.get(variableName);
                 checkTypeAndThrowError(tomlValue.kind(), variableName, effectiveType);
                 value = retrieveRecordValues((TomlTableNode) tomlValue, variableName, type);
                 break;
             case TypeTags.TABLE_TAG:
-                tomlValue = valueMap.get(variableName);
                 checkTypeAndThrowError(tomlValue.kind(), variableName, effectiveType);
                 value = retrieveTableValues((TomlTableArrayNode) tomlValue, variableName, (TableType) effectiveType);
                 break;
             default:
-                throw new TomlException(String.format(CONFIGURATION_NOT_SUPPORTED, effectiveType.toString()));
+                throw new TomlException(
+                        String.format(CONFIGURATION_NOT_SUPPORTED, effectiveType.toString(), variableName));
         }
         return value;
     }
 
     private static Object retrievePrimitiveValue(TomlNode tomlValue, String variableName, Type type) {
+        checkTomlValueStructure(tomlValue, TomlType.KEY_VALUE, variableName);
+        tomlValue = ((TomlKeyValueNode) tomlValue).value();
         checkTypeAndThrowError(tomlValue.kind(), variableName, type);
         return getBalValue(variableName, type.getTag(), ((TomlBasicValueNode<?>) tomlValue).getValue());
     }
 
-    private static Object retrieveArrayValues(TomlArrayValueNode arrayNode, String variableName,
+    private static Object retrieveArrayValues(TomlNode tomlValue, String variableName,
                                               ArrayType effectiveType) {
+        checkTomlValueStructure(tomlValue, TomlType.KEY_VALUE, variableName);
+        tomlValue = ((TomlKeyValueNode) tomlValue).value();
+        checkTypeAndThrowError(tomlValue.kind(), variableName, effectiveType);
         Type elementType = effectiveType.getElementType();
-        List<TomlValueNode> arrayList = arrayNode.elements();
-        checkTypeAndThrowError(arrayList.get(0).kind(), variableName, elementType);
-        return new ArrayValueImpl(effectiveType, arrayList.size(), createArray(variableName, arrayList,
-                                                                               elementType.getTag()));
+        List<TomlValueNode> arrayList = ((TomlArrayValueNode) tomlValue).elements();
+        if (!isPrimitiveType(elementType.getTag())) {
+            //Remove after supporting all arrays
+            throw new TomlException(String.format(CONFIGURATION_NOT_SUPPORTED, effectiveType.toString(), variableName));
+        }
+        return new ArrayValueImpl(effectiveType, arrayList.size(), createArray(variableName, arrayList, elementType));
+    }
+
+    private static void checkTomlValueStructure(TomlNode tomlValue, TomlType expectedType, String variableName) {
+        if (tomlValue.kind() != expectedType) {
+            throw new TomlException(String.format(INVALID_TOML_FILE + INVALID_TOML_STRUCTURE, variableName,
```

Comment:
Shall we add the toml line numbers for errors as well.

Reply:
Created a issue https://github.com/ballerina-platform/ballerina-lang/issues/28711